### PR TITLE
Implemented store level operations for the updated DLC

### PR DESCRIPTION
**Pull request by sdkottegoda**
This addresses the jira https://wso2.org/jira/browse/MB-1144
Please merge https://github.com/wso2/carbon-business-messaging/pull/92 along with this or else the tests will fail.

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessageStore.java [hunk 1]
```diff
@@ -222,12 +298,22 @@ public void addMessageToExpiryQueue(Long messageId, Long expirationTime,
...
-    public void deleteAllMessageMetadata(String storageQueueName) throws AndesException;
+    public int deleteAllMessageMetadata(String storageQueueName) throws AndesException;
+
+    /**
+     * Store level method to remove all metadata in a dead letter channel
+     *
+     * @param dlcQueueName name of the queue being purged
+     * @throws AndesException
+     */
+    public int clearDlcQueue(String dlcQueueName) throws AndesException;
```
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessageStore.java]: In some method signatures "Dlc" is used as "DLC", shall we use only one of those?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/metrics/MetricsConstants.java [hunk 2]
```diff
@@ -83,9 +87,18 @@
      */
     public static final String GET_NEXT_MESSAGE_METADATA_FROM_QUEUE = PREFIX + "store.nextMessageMetadataFromQueue.get";
     /**
-     * Delete message part
+     * Get next message meta data in dead letter channel for queue
+     */
+    public static final String GET_NEXT_MESSAGE_METADATA_IN_DLC_FOR_QUEUE = PREFIX + "store"
```
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/metrics/MetricsConstants.java]: Can we get "store" to next line?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java [hunk 3]
```diff
@@ -30,24 +30,12 @@
...
-import org.wso2.andes.kernel.AndesMessage;
-import org.wso2.andes.kernel.AndesMessageMetadata;
-import org.wso2.andes.kernel.AndesMessagePart;
-import org.wso2.andes.kernel.AndesRemovableMetadata;
-import org.wso2.andes.kernel.AndesSubscription;
-import org.wso2.andes.kernel.AndesUtils;
-import org.wso2.andes.kernel.DisablePubAckImpl;
-import org.wso2.andes.kernel.FlowControlListener;
-import org.wso2.andes.kernel.MessagingEngine;
+import org.wso2.andes.kernel.*;
```
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java]: Expand imports.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java [hunk 4]
```diff
@@ -461,154 +449,49 @@ public void restoreMessagesFromDeadLetterQueue(@MBeanOperationParameter(name = "
...
-        return compositeDataList.toArray(new CompositeData[compositeDataList.size()]);
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public long getNumberMessagesInDLCForQueue(String queueName) {
-        return 0;
+    public long getNumberMessagesInDLCForQueue(String queueName) throws MBeanException{
```
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java]: NumberMessages or NumberOfMessages?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java [hunk 5]
```diff
@@ -882,25 +783,21 @@ private String getReadableNameForMessageContentType(String contentType) {
...
 
         long messageCount = 0;
         try {
-        if (msgPattern.equals("queue")) {
-            messageCount = MessagingEngine.getInstance().getMessageCountOfQueue(queueName);
-        } else if (msgPattern.equals("topic")) {
-            //TODO: hasitha - to implement
-        }
+            if (!DLCQueueUtils.isDeadLetterQueue(queueName)) {
+                if (msgPattern.equals("queue")) {
```
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java]: Let's flip msgPattern.equals("queue").

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java [hunk 6]
```diff
@@ -527,6 +528,22 @@ public void moveMetadataToQueue(long messageId,
      * {@inheritDoc}
      */
     @Override
+    public void moveMetadataToDLC(long messageId, String dlcQueueName) throws AndesException {
+        throw new NotImplementedException();
```
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java]: Better using UnsupportedException.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSConstants.java [hunk 7]
```diff
@@ -167,8 +168,9 @@
             "INSERT INTO " + METADATA_TABLE + " (" +
                     MESSAGE_ID + "," +
                     QUEUE_ID + "," +
-                    METADATA + ")" +
-                    " VALUES ( ?,?,? )";
+                    DLC_QUEUE_ID + "," +
+                    METADATA + ") " +
+                    "VALUES ( ?,?,-1,? )";
```
Comment (AnujaLK) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSConstants.java]: Add a comment describing why we use "-1" here

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java (hunk 6 above)
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java]: I think its high time to remove cassandra...
I'm gonna do it.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessageStore.java (hunk 1 above)
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessageStore.java]: Changed all the places to have "DLC". Fixed in the PR https://github.com/wso2/andes/pull/381

File: modules/andes-core/broker/src/main/java/org/wso2/andes/metrics/MetricsConstants.java (hunk 2 above)
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/metrics/MetricsConstants.java]: Fixed in https://github.com/wso2/andes/pull/381

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java (hunk 3 above)
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java]: Exported in https://github.com/wso2/andes/pull/381

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java (hunk 4 above)
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java]: It should be NumberOfMessages. Fixed in https://github.com/wso2/andes/pull/381

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java (hunk 5 above)
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java]: Flipped in https://github.com/wso2/andes/pull/381

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java (hunk 6 above)
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/CQLBasedMessageStoreImpl.java]: Cassandra is removed and this class isn't present anymore.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSConstants.java (hunk 7 above)
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSConstants.java]: Added a comment in https://github.com/wso2/andes/pull/381

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/HectorBasedMessageStoreImpl.java [hunk 8]
```diff
@@ -356,6 +358,22 @@ public void moveMetadataToQueue(long messageId, String currentQueueName,
      * {@inheritDoc}
      */
     @Override
+    public void moveMetadataToDLC(long messageId, String dlcQueueName) throws AndesException {
+        throw new NotImplementedException();
```
Comment (pumudu88) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/cassandra/HectorBasedMessageStoreImpl.java]: HectorBasedMessageStoreImpl / CQLBasedMessageStoreImpl implementations removed from MB.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java [hunk 9]
```diff
@@ -921,4 +818,133 @@ public int getSubscriptionCount( String queueName){
         }
     }
 
+    /**
+     * Method to display a list of messages when browsed
+     */
+    private CompositeData[] getDisplayableMetaData(List<AndesMessageMetadata> metadataList) throws MBeanException{
+        List<CompositeData> compositeDataList = new ArrayList<>();
+        try {
+            for (AndesMessageMetadata andesMessageMetadata : metadataList) {
```
Comment (kasun04) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java]: Isn't the following blog too long. Can we split this in to private methods?

Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/information/management/QueueManagementInformationMBean.java]: Divided into three methods by the PR https://github.com/wso2/andes/pull/403/files

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/Andes.java [hunk 10]
```diff
@@ -438,6 +439,17 @@ public AndesMessagePart getContent(long messageId, int offsetValue) throws Andes
     }
 
     /**
+     * Get content chunks for a list of message ids from store
```
Comment (lafernando) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/Andes.java]: First Javadoc sentence should end with a period, fix in all the places.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java [hunk 11]
```diff
@@ -847,6 +1017,113 @@ public AndesMessageMetadata getMetadata(long messageId) throws AndesException {
...
+            preparedStatement = connection
+                    .prepareStatement(RDBMSConstants.PS_SELECT_METADATA_IN_DLC_FOR_QUEUE);
+            preparedStatement.setLong(1, firstMsgId - 1);
+            preparedStatement.setInt(2, getCachedQueueID(storageQueueName));
+            preparedStatement.setInt(3, getCachedQueueID(dlcQueueName));
+
+            results = preparedStatement.executeQuery();
+            int resultCount = 0;
+            while (results.next()) {
+
```
Comment (lafernando) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java]: Remove unnecessary whitespace.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java [hunk 12]
```diff
@@ -1323,6 +1601,44 @@ public void deleteAllMessageMetadata(String storageQueueName) throws AndesExcept
...
+            return deletedMessagecount;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public int clearDlcQueue(String dlcQueueName) throws AndesException {
+
```
Comment (lafernando) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java]: Remove unnecessary whitespace, fix in all the places.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/Andes.java (hunk 10 above)
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/Andes.java]: Addressed by the PR https://github.com/wso2/andes/pull/403

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java (hunk 11 above)
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java]: Addressed by the PR https://github.com/wso2/andes/pull/403

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java (hunk 12 above)
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java]: Fixed in this class and some other identified classes by the PR https://github.com/wso2/andes/pull/403. Will progress onto the rest of the classes as well

File: modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java [hunk 13]
```diff
@@ -793,6 +859,110 @@ public AndesMessageMetadata getMetadata(long messageId) throws AndesException {
...
+                        + "\nresponse: metadata count " + metadataList.size());
+            }
+        } catch (SQLException e) {
+            throw rdbmsStoreUtils.convertSQLException("Error occurred while retrieving messages between msg id " +
+                    firstMsgId + " and " + lastMsgId + "in DLC:" + dlcQueueName + " from queue " + storageQueueName, e);
+        } finally {
+            metaListRetrievalContext.stop();
+            contextRead.stop();
+            String task = RDBMSConstants.TASK_RETRIEVING_METADATA_RANGE_IN_DLC_FROM_QUEUE + storageQueueName;
+            close(resultSet, task);
```
Comment (dinusha-dilrukshi) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java]: It would be nice if you could introduce a new wrapper method like "closeAllConnections(connection, preparedStatement, resultSet) {}" and call that method everywhere that you need to close connection, preparedstatement and resultset instead of calling close() method three times.

Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/store/rdbms/RDBMSMessageStoreImpl.java]: Introduced two new methods to for close(connection, prepared statement) and close(connection, prepared statement, result set) to avoid repetition. Addressed by the PR https://github.com/wso2/andes/pull/403
